### ISSUE TITLE
Unexpected error message when using tagged union in an nested object.

### ISSUE THREAD

**Issue by pissang** [microsoft/TypeScript#46374]
# Bug Report

### 🔎 Search Terms

+ Tagged Union
+ Type Narrowing
+ Nested Object

### 🕗 Version & Regression Information

- This changed between versions ___3.9___ and ___4.5-beta___

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgDKgnKBxKcAmwE4A8gEYBWECYyA3gFDLPJgCeADhAFzIDkAGwxY+DJiwAevEAFcAtmWjjmbafMVRlyCQCY1CpS2Rs9yWQc0BfMaEixEKAEoFgcAbhfEw5KjXpb2Ll4+PEI3US0pM3VDFlVoiy0ofQ0GawZAlA9CLx9qWgBeNGEcUKJSSnzkAB9kZzD3MtzKmgBuMUzkAGV2ARQixiMEAHsBYeTkbPLvFtpagGcwKFAAczSxBgB6TeQAUSgocd4AFU4UPjpWM94AIiEQTCgb1uRLPmRgeeQ5T-nV1gAFigYKMxgB3f4cQ5cKBgIhfGCHORXLj8dAPLBTZq+MB8XgSAA0xiJuiJJgYIxAi2QIzGEyxFRxyAGAWu-Huj1ElnalOpizYfV4PQF-X8Q1GRzFRhY2z2B0lp1RfDuJRu70+ZmGtDg8z+KxAcDIfVYwxR5xuZTcaoAdAAKHQAZh0OgAlFojJlghyRFprNyxEA)

### 💻 Code

```ts
interface LinearGradientObject {
    type: 'linear'

    x: number
    y: number
    x2: number
    y2: number
}

interface RadialGradientObject {
    type: 'radial'

    x: number
    y: number
    r: number
}

type GradientObject = LinearGradientObject | RadialGradientObject;

type Style = {
    color: GradientObject | string
}


const style: Style = {
    color: {
        // Error: Type '"linear"' is not assignable to type '"radial"'.(2322)
        // It's not expected.
        type: 'linear'
    }
};

// Error: Type '{ type: "linear"; }' is missing the following properties from type 'LinearGradientObject': x, y, x2, y2
// It's expected
const color: GradientObject = {
    type: 'linear'
};

```

### 🙁 Actual behavior

The error when assigning the literal nested object to the `style` variable in the code above is confusing. 
```
Error: Type '"linear"' is not assignable to type '"radial"'.(2322)
```

### 🙂 Expected behavior

It will be much clearer if the error message is:
```
Error: Type '{ type: "linear"; }' is missing the following properties from type 'LinearGradientObject': x, y, x2, y2
```

**Comment by fatcerberus** [comment]
> you are setting color to GradientObject, which is discriminated by type: 'linear' and therefore expects the missing properties from LinearGradientObject

That’s what’s *supposed* to happen, but according to the OP the actual error produced is instead:

```
Error: Type '"linear"' is not assignable to type '"radial"'.
```

**Comment by jcalz** [comment]
I'm guessing this was caused by #37589?

**Comment by RyanCavanaugh** [comment]
This now reports the missing `LinearGradientObject` properties for the nested `color` object rather than reporting that `"linear"` is not assignable to `"radial"`.

The change is present from TypeScript 5.1.0-dev.20230413 and was introduced by [#53709](https://github.com/microsoft/TypeScript/pull/53709). TypeScript 6.0.3 and the current native build retain the improved diagnostic.
